daemon/main.vala

```diff
@@ -0,0 +1,150 @@
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+namespace GalaDaemon
+{
+	[DBus (name = "org.gnome.SessionManager")]
+	public interface SessionManager : Object
+	{
+		public abstract async ObjectPath RegisterClient (string app_id, string client_start_id) throws IOError;
+	}
+
+	[DBus (name = "org.gnome.SessionManager.ClientPrivate")]
+	public interface SessionClient : Object
+	{
+		public abstract void EndSessionResponse (bool is_ok, string reason) throws IOError;
+
+		public signal void Stop () ;
+		public signal void QueryEndSession (uint flags);
+		public signal void EndSession (uint flags);
+		public signal void CancelEndSession ();
+	}
+
+	public class Daemon
+	{
+		SessionClient? sclient = null;
+
+		public Daemon ()
+		{
+			register.begin ((o, res)=> {
+				bool success = register.end (res);
+				if (!success) {
+					message ("Failed to register with Session manager");
+				}
+			});
+
+			var menu_daemon = new MenuDaemon ();
+			menu_daemon.setup_dbus ();
+		}
+
+		public static async SessionClient? register_with_session (string app_id)
+		{
+			ObjectPath? path = null;
+			string? msg = null;
+			string? start_id = null;
+
+			SessionManager? session = null;
+			SessionClient? session_client = null;
+
+			start_id = Environment.get_variable ("DESKTOP_AUTOSTART_ID");
+			if (start_id != null) {
+				Environment.unset_variable ("DESKTOP_AUTOSTART_ID");
+			} else {
+				start_id = "";
+				warning ("DESKTOP_AUTOSTART_ID not set, session registration may be broken (not running via session?)");
+			}
+
+			try {
+				session = yield Bus.get_proxy (BusType.SESSION, "org.gnome.SessionManager", "/org/gnome/SessionManager");
+			} catch (Error e) {
+				warning ("Unable to connect to session manager: %s", e.message);
+				return null;
+			}
+
+			try {
+				path = yield session.RegisterClient (app_id, start_id);
+			} catch (Error e) {
+				msg = e.message;
+				warning ("Error registering with session manager: %s", e.message);
+				return null;
+			}
+
+			try {
+				session_client = yield Bus.get_proxy (BusType.SESSION, "org.gnome.SessionManager", path);
+			} catch (Error e) {
+				warning ("Unable to get private sessions client proxy: %s", e.message);
+				return null;
+			}
+
+			return session_client;
+		}
+
+		private async bool register ()
+		{
+			sclient = yield register_with_session ("io.elementary.gala.daemon");
+
+			sclient.QueryEndSession.connect (()=> {
+				end_session (false);
+			});
+
+			sclient.EndSession.connect (()=> {
+				end_session (false);
+			});
+
+			sclient.Stop.connect (()=> {
+				end_session (true);
+			});
+
+			return true;
+		}
+
+		private void end_session (bool quit)
+		{
+			if (quit) {
+				Gtk.main_quit ();
+				return;
```

Comment:
Is there a reason why we return here? Does GNOME SessionManager automatically ends the session for this process when it quits?

Reply:
Yes, session manager sees the disconnection in DBus. Besides, even if we don't return here and try to respond to the session manager, it won't work since there isn't a Gtk main loop to communicate with DBus anymore.